src/java/org/apache/cassandra/index/sai/iterators/HnswIntersectionIterator.java

```diff
@@ -0,0 +1,241 @@
+
+/**
+ * The HnswIntersectionIterator is an implementation of the KeyRangeIterator
+ * that handles intersection operations between normal SAI predicates and Hierarchical
+ * Navigable Small World (HNSW) graphs.
+ *
+ * HNSW is special because it returns results in similarity order, not in key order,
+ * and because it's really more of an order by than a where clause; HNSW doesn't
+ * "eliminate" any keys, they will all be in the results if you iterate far down enough.
+ *
+ * HnswIntersectionIterator performs what the Pinterest guys call Post Filtering:
+ * we perform the HNSW search, then go through the keys from the other predicates
+ * looking for matches.  If we don't get enough matches, we expand the HNSW search
+ * window and perform the search over again.
+ *
+ * To attempt to minimize the passes, we double the HNSW window with each iteration.
+ */
+public class HnswIntersectionIterator extends KeyRangeIterator
+{
+    private final Builder builder;
+    private final KeyRangeIterator hnswIterator;
+    private HnswOnePassIterator onePassIterator;
+    private final Set<PrimaryKey> hnswMatches = new HashSet<>();
+    private final Set<PrimaryKey> seenMatchingKeys = new HashSet<>();
+    private int hnswBatchSize;
+
+    protected HnswIntersectionIterator(Builder builder)
+    {
+        super(builder.statistics);
+        this.builder = builder;
+        this.hnswIterator = builder.hnswIterator;
+        hnswBatchSize = limit();
+        this.onePassIterator = newOnePassIterator(builder);
+        loadMoreHnswMatches();
+    }
+
+    private HnswOnePassIterator newOnePassIterator(Builder builder)
+    {
+        return new HnswOnePassIterator(builder.limit,
+                                       builder.otherSuppliers.stream().map(Supplier::get).collect(Collectors.toList()));
+    }
+
+    @Override
+    public PrimaryKey skipTo(PrimaryKey nextKey)
+    {
+        return onePassIterator.skipTo(nextKey);
+    }
+
+    @Override
+    protected void performSkipTo(PrimaryKey nextKey)
+    {
+        // external callers will call skipTo()
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void close() throws IOException
+    {
+        onePassIterator.close();
+    }
+
+    // TODO do we need to compute all the rows, and then re-sort by PK?
+    @Override
+    protected PrimaryKey computeNext()
+    {
+        // TODO integrate this better with the actual HNSW batch window
+        while (seenMatchingKeys.size() < limit())
+        {
+            while (onePassIterator.hasNext())
+            {
+                var next = onePassIterator.next();
+                if (seenMatchingKeys.add(next))
+                {
+                    return next;
+                }
+            }
+
+            if (!loadMoreHnswMatches()) {
+                return endOfData();
+            } else {
+                hnswBatchSize *= 2;
+            }
+            onePassIterator = newOnePassIterator(builder);
```

Comment:
Is this expecting to reset the `onePassIterator`? If so, I'm not sure it will work. The underlying iterators are single pass so will not reset. Am I missing something here?

Reply:
It rebuilds those from the Supplier that the builder saves